finality-aleph/src/data_io/data_store.rs

```diff
@@ -0,0 +1,636 @@
+            match status {
+                PendingTopBlock => {
+                    self.pending_proposals
+                        .insert(proposal.clone(), PendingProposalInfo::new(status));
+                    self.register_block_import_trigger(proposal, &proposal.top_block());
+                    self.register_next_finality_trigger(proposal);
+                }
+                TopBlockImportedButIncorrectBranch => {
+                    self.pending_proposals
+                        .insert(proposal.clone(), PendingProposalInfo::new(status));
+                    self.register_next_finality_trigger(proposal);
+                }
+                TopBlockImportedButNotFinalizedAncestor => {
+                    self.pending_proposals
+                        .insert(proposal.clone(), PendingProposalInfo::new(status));
+                    self.register_next_finality_trigger(proposal);
+                }
+                Available(_) => {
+                    // Proposal available, no need to register any dependencies
+                    return;
+                }
+            }
+        }
+        // This line is reached only if the proposal is not available
+        let proposal_info = self
+            .pending_proposals
+            .get_mut(proposal)
+            .expect("exists as checked above");
+        proposal_info.messages.insert(id);
+        message_info.pending_proposals.insert(proposal.clone());
+    }
+
+    fn on_message_dependencies_resolved(&self, message: Message) {
+        trace!(target: "afa", "Sending message from DataStore {:?}", message);
+        if let Err(e) = self.messages_for_aleph.unbounded_send(message) {
+            error!(target: "afa", "Unable to send a ready message from DataStore {}", e);
+        }
+    }
+
+    fn assign_fresh_message_id(&mut self) -> MessageId {
+        self.next_free_id += 1;
+        self.next_free_id - 1
+    }
+
+    // This is called upon a proposal being available -- we remove it from the set of
+    // proposals a message waits for.
+    fn remove_proposal_from_pending_message(&mut self, proposal: &AlephProposal<B>, id: MessageId) {
+        let mut message_info = match self.pending_messages.remove(&id) {
+            Some(message_info) => message_info,
+            None => {
+                warn!(target: "afa", "Message {:?} not found when resolving a proposal dependency {:?}.", id, proposal);
+                return;
+            }
+        };
+        message_info.pending_proposals.remove(proposal);
+        if message_info.pending_proposals.is_empty() {
+            self.on_message_dependencies_resolved(message_info.message);
+        } else {
+            // We reinsert the message because it still has pending proposals.
+            self.pending_messages.insert(id, message_info);
+        }
+    }
+
+    fn remove_id_from_pending_proposal(&mut self, proposal: &AlephProposal<B>, id: MessageId) {
```

Comment:
I would rename this method to something like `remove_message_id_from_pending_proposal` or `remove_message_dependency_for_pending_proposal`